Prefetch glossary + prerendering (#27589)
mdn/content: files/en-us/glossary/prefetch/index.md

{{GlossarySidebar}}

Prefetching refers to the practice of speculatively fetching resources in the background for pages that the user is _likely_ to navigate to in the near future.
This can significantly reduce the load time for the prefetched page if the user does choose to navigate to it.
Prefetching might be used, for example, to fetch the page resources linked by a "Next" button, or a link popup that a user hovers over, or search results.

### Resource prefetching

Resources should be prefetched based on how likely it is that they will be needed in a future navigation.
Browsers can infer this automatically for some resources, such as the current URL in the address bar.

Browsers will also prefetch content for [`<link>`](/en-US/docs/Web/HTML/Element/link) elements that have the [`rel="prefetch"`](/en-US/docs/Web/HTML/Attributes/rel/prefetch) attribute.
This allows developers to hint to the browser the likely navigation from the current page.
Note that only these links are prefetched (so `<a>` elements are not), and that they are fetched at lower priority that resources used by the current page, including elements with `fetchPriority="low"`.

> 

Author: @noamr Note, I wasn't sure exactly what you meant by

> _This is a common use of prefetching - loading "app" resources from a smaller landing page."_

Do you mean this?:

```suggestion

Prefetching can be used to fetch both HTML and sub-resources for a possible next navigation.  A common use case is to have a simple website landing page that fetches more "heavy-weight" resources used by the rest of the site.

```

Reviewer: Yes exactly!

Author: Thanks! You've put a lot of work getting this feature in - I'm in awe at how many places you've had to touch to get this into the spec and various browsers!!!

{{GlossarySidebar}}

Prefetching refers to the practice of speculatively fetching resources in the background for pages that the user is _likely_ to navigate to in the near future.
This can significantly reduce the load time for the prefetched page if the user does choose to navigate to it.
Prefetching might be used, for example, to fetch the page resources linked by a "Next" button, or a link popup that a user hovers over, or search results.

> Prefetching is the first step in {{Glossary("prerender","prerendering")}} and has the same purpose: to reduce time to display for the likely next navigation.

Reviewer: Not sure if it's the first step in prerender, this feels confusing

Author: Fetching pages ahead of time kind of has to be the first step in prerendering. But since you're the expert and this can be remove without harm I will do so.

Reviewer: It is the first step in prerendering, but it's also the first step in a regular navigation... nothing special about prerendering in that regard

> It differs from [preloading](/en-US/docs/Web/HTML/Attributes/rel/preload), which is the term used to refer to controlling the fetch priority for resources in the current page.

Reviewer: Preloading is about loading a subresource to be used in the current page, "controlling the fetch priority" feels confusing

Author: FYI only, 
- This paragraph was there to provide a very succinct overview to differentiate prefetch, preload, prerender. I've deleted this since it wasn't "necessary". 
- As I understand it, preloading is about getting resources you need for the current page earlier than you would otherwise - hence "controlling the fetch priority".

Resources should be prefetched based on how likely it is that they will be needed in a future navigation.
Browsers can infer this automatically for some resources, such as the current URL in the address bar.

Browsers will also prefetch content for [`<link>`](/en-US/docs/Web/HTML/Element/link) elements that have the [`rel="prefetch"`](/en-US/docs/Web/HTML/Attributes/rel/prefetch) attribute.
This allows developers to hint to the browser the likely navigation from the current page.
Note that only these links are prefetched (so `<a>` elements are not), and that they are fetched at lower priority that resources used by the current page, including elements with `fetchPriority="low"`.

> Prefetching can be used to fetch both HTML and sub-resources for a possible next navigation.  A common use case is to have a simple website landing page that fetches more "heavy-weight" resources used by the rest of the site.

Reviewer 2: ```suggestion
Prefetching can be used to fetch both HTML and sub-resources for a possible next navigation. A common use case is to have a simple website landing page that fetches more "heavy-weight" resources used by the rest of the site.
```

Reviewer 2: Passing linter

> Note that only these links are prefetched (so `<a>` elements are not), and that they are fetched at lower priority that resources used by the current page.

Reviewer: ... event lower than elements with fetchpriority=low

Author: Thanks.

```suggestion
Note that only these links are prefetched (so `<a>` elements are not), and that they are fetched at lower priority that resources used by the current page, including elements with `fetchPriority="low"`.
```

```html

> <link rel="prefetch" href="/app/style.css" />
> <link rel="prefetch" href="https://example.com/landing-page" />

Reviewer 2: ```suggestion
<link rel="prefetch" href="/app/style.css">
<link rel="prefetch" href="https://example.com/landing-page">
```

> <link rel="prefetch" href="https://www.example.com/solutions" />

Reviewer: Maybe also have an example that loads a subresource? e.g. `<link rel="prefetch" href="/app/style.css">` `https://example.com/landing-page`. This is a common use of prefetching - loading "app" resources from a smaller landing page.

Author: Sounds good. 
```suggestion
<link rel="prefetch" href="/app/style.css"> 
<link rel="prefetch" href="https://example.com/landing-page">
```

```

The fetch request for a `prefetch` operation results an HTTP Request that includes the HTTP header [`Sec-Purpose: prefetch`](/en-US/docs/Web/HTTP/Headers/Sec-Purpose), which a server might use to change the cache timeouts for the resources, or perform other special handling.
The request should also include the {{HTTPHeader("Sec-Fetch-Dest")}} header with the value set to `empty`.
The {{HTTPHeader("Accept")}} header in the request should match the value used for normal navigation requests, so that the browser can find the matching cached resources following navigation.
If a response is returned, it is cached with the request in the HTTP cache.

> > **Note:** Browser vendors are currently aligning around the behaviour described here, and may use different headers.

Author: ```suggestion
> **Note:** Browser vendors are currently aligning around the HTTP headers that are sent, and may use different headers and values.
```

> The best place to track specification compliance is the [`Sec-Purpose`](/en-US/docs/Web/HTTP/Headers/Sec-Purpose) compatibility table.

### DNS Prefetching

[DNS prefetching](/en-US/docs/Web/HTML/Attributes/rel/dns-prefetch) resolves domain names in advance thereby speeding up load times by reducing the time associated with domain lookup at request time.

```html

> <link rel="dns-prefetch" href="https://example.com/" />

Reviewer 2: ```suggestion
<link rel="dns-prefetch" href="https://example.com/">
```
